### PR TITLE
8078471: Backspace does not work in JFileChooser with GTK L&F

### PR DESCRIPTION
**Pull request by kumarabhi006**
The Backspace key doesn't lead to parent directory of current directory for JFileChooser in GTK LAF. 
Added the lazy input value in `GTKLookandFeel` and fix is working fine.

Implemented a manual test to check BackSpace behavior for all installed LAF in Linux platform.
The test mentioned in [JDK-8078471](https://bugs.openjdk.org/browse/JDK-8078471) `javax/swing/JFileChooser/4150029/bug4150029.html` also working fine with GTK option (-client -Dswing.defaultlaf=com.sun.java.swing.plaf.gtk.GTKLookAndFeel).

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8078471](https://bugs.openjdk.org/browse/JDK-8078471): Backspace does not work in JFileChooser with GTK L&F

### Reviewers
 * [Prasanta Sadhukhan](https://openjdk.org/census#psadhukhan) (@prsadhuk - **Reviewer**) ⚠️ Review applies to [3b19e5cd](https://git.openjdk.org/jdk/pull/11291/files/3b19e5cd58e2d3fa619e01b90bf3c18ee8293f5a)
 * [Phil Race](https://openjdk.org/census#prr) (@prrace - **Reviewer**) ⚠️ Review applies to [3b19e5cd](https://git.openjdk.org/jdk/pull/11291/files/3b19e5cd58e2d3fa619e01b90bf3c18ee8293f5a)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11291/head:pull/11291` \
`$ git checkout pull/11291`

Update a local copy of the PR: \
`$ git checkout pull/11291` \
`$ git pull https://git.openjdk.org/jdk pull/11291/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11291`

View PR using the GUI difftool: \
`$ git pr show -t 11291`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11291.diff">https://git.openjdk.org/jdk/pull/11291.diff</a>

</details>

### CODE REVIEW

**Comment by kumarabhi006** [comment]
If JFileChooser current directory is set to root directory then pressing BackSpace key changes to user's home directory in GTK LAF which is incorrect. It should remain at root directory. 
Planning to raise a new bug for this.

**Comment by kumarabhi006** [comment]
> If JFileChooser current directory is set to root directory then pressing BackSpace key changes to user's home directory in GTK LAF which is incorrect. It should remain at root directory. Planning to raise a new bug for this.

@prsadhuk @jayathirthrao As discussed in the meeting, will not raise a new bug and update this PR accordingly.

**Comment by kumarabhi006** [comment]
Updated to handle backspace key at root directory.

**Comment by kumarabhi006** [comment]
Run the CI clienlibs job with or without fix. There were few failures observed.
2 of them are common in both and 1 failure is observed in particular machine.

Link is added in JBS.

**Comment by prsadhuk** [comment]
I wonder how it works in Nimbus as this mapping is not there in Synth/NimbusLookAndFeel and still it works and both GTK/Nimbus extends Synth. Can you please check?

> If JFileChooser current directory is set to root directory then pressing BackSpace key changes to user's home directory in GTK LAF which is incorrect. It should remain at root directory.

You have a similar fix for GTK https://github.com/openjdk/jdk/commit/9d0009e92b790b43153e3db353db775e6ff731cb 
Can it be  reused in this case also?

Also, is it not possible to create automated test, like calling JFileChooser.getCurrentDirectory() and fail if it's not the parent directory (or if it still remains in present directory) after BACKSPACE is pressed?

**Comment by kumarabhi006** [comment]
> I wonder how it works in Nimbus as this mapping is not there in Synth/NimbusLookAndFeel and still it works and both GTK/Nimbus extends Synth. Can you please check?


 There is a mapping for squarebutton in skin.laf `<uiProperty name="squareButton" type="BOOLEAN" value="false"/>`. 
Setting it true is having similar effect in Nimbus LAF also. So, I think in Nimbus LAF the squarebutton property is set through skin.laf file.

> You have a similar fix for GTK https://github.com/openjdk/jdk/commit/9d0009e92b790b43153e3db353db775e6ff731cb
Can it be reused in this case also?

Don't think it can be reused in this case as the event is triggered from either double click or keyboard (Enter) on "../" directory present in JList. 
I will verify this once.

> Also, is it not possible to create automated test, like calling JFileChooser.getCurrentDirectory() and fail if it's not the parent directory (or if it still remains in present directory) after BACKSPACE is pressed?

I tried implementing an automated test case but in GTK LAF the cursor is focused on textfield and the backspace event is not triggered to change directory.

**Comment by prsadhuk** [comment]
> There is a mapping for squarebutton in skin.laf `<uiProperty name="squareButton" type="BOOLEAN" value="false"/>`.
> Setting it true is having similar effect in Nimbus LAF also. So, I think in Nimbus LAF the squarebutton property is set through skin.laf file.

This mapping ""BACK_SPACE", "Go Up"," I guess has nothing to do with squarebutton property..Probably you are confusing it with your other fix?

> I tried implementing an automated test case but in GTK LAF the cursor is focused on textfield and the backspace event is not triggered to change directory.

You probably can do requestFocusInWindow to proper component...you can try to find the who has the focus by getFocusOwner() when you try to do manually BACKSPACE and try to set to that in automated test...

**Comment by kumarabhi006** [comment]
> This mapping ""BACK_SPACE", "Go Up"," I guess has nothing to do with squarebutton property..Probably you are confusing it with your other fix?

Yeah, I am sorry. I will check in Nimbus then how it's happening.

> You probably can do requestFocusInWindow to proper component...you can try to find the who has the focus by getFocusOwner() when you try to do manually BACKSPACE and try to set to that in automated test...

Ok.

**Comment by kumarabhi006** [comment]
> Also, is it not possible to create automated test, like calling JFileChooser.getCurrentDirectory() and fail if it's not the parent directory (or if it still remains in present directory) after BACKSPACE is pressed?

Added Automated test case. Tested in CI pipeline, Link is attached in JBS.

**Comment by kumarabhi006** [comment]
> I wonder how it works in Nimbus as this mapping is not there in Synth/NimbusLookAndFeel and still it works and both GTK/Nimbus extends Synth. Can you please check?

The " Go Up" action in Nimbus is mapped through FilePane class.
But how Back_Space is mapped to "Go Up" is not clear.
There is a mapping in SynthStyle "BACK_SPACE", "Go Up", but even after commenting the backspace works fine.

**Review by prsadhuk — APPROVED** [review]

File: test/jdk/com/sun/java/swing/plaf/gtk/TestBackSpaceAction.java [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+                tmpDir = System.getProperty("user.home");
+            }
+            testDir = new File(tmpDir, "testDir");
+            testDir.mkdir();
+            testDir.deleteOnExit();
+
+            // create sub directory inside test directory
+            subDir = new File(testDir, "subDir");
+            subDir.mkdir();
+            subDir.deleteOnExit();
```
Comment (prrace) [inline test/jdk/com/sun/java/swing/plaf/gtk/TestBackSpaceAction.java:75]: To my surprise, this should be OK. The spec says
"Files (or directories) are deleted in the reverse order that they are registered."

**Review by prrace — APPROVED** [review]
Please split the long line before pushing.
Also you'll probably need to update the PR to match the new bug summary.
ALWAYS consider if a bug summary is well written as part of the bug evaluation

**Comment by kumarabhi006** [comment]
> Please split the long line before pushing. Also you'll probably need to update the PR to match the new bug summary. ALWAYS consider if a bug summary is well written as part of the bug evaluation

PR description changed to match the bug summary and long line split taken care.

**Pull request closed without merging**